applications/hotornot/src/hon_trie.erl

```diff
@@ -0,0 +1,154 @@
+%%%-------------------------------------------------------------------
+%%% @copyright (C) 2011-2016, 2600Hz INC
+%%% @doc
+%%% Rater whapp; send me a DID, get a rate back
+%%%
+%%% (Words * Bytes/Word) div (Prefixes) ~= Bytes per Prefix
+%%% (1127494 * 8) div 78009 = 115
+%%%
+%%% Processing:
+%%% timer:tc(hon_trie, match_did, [<<"53341341354">>]).
+%%%  {132,...}
+%%% timer:tc(hon_util, candidate_rates, [<<"53341341354">>]).
+%%%  {16989,...}
+%%% @end
+%%% @contributors
+%%%   James Aimonetti
+%%%-------------------------------------------------------------------
+-module(hon_trie).
+-behaviour(gen_server).
+
+-export([start_link/0
+        ,match_did/1
+        ]).
+
+-export([init/1
+        ,handle_call/3
+        ,handle_cast/2
+        ,handle_info/2
+        ,terminate/2
+        ,code_change/3
+        ]).
+
+-export([build_trie/1, build_trie/2]).
+
+-include("hotornot.hrl").
+
+-type state() :: trie:trie() | pid_ref().
+
+-spec start_link() -> {'ok', pid()}.
+start_link() ->
+    gen_server:start_link({'local', ?MODULE}, ?MODULE, [], []).
+
+-type match_return() :: {'error', 'not_found'} |
+                        {'ok', {string(), ne_binaries()}}.
+-spec match_did(ne_binary()) -> {'ok', kz_json:objects()}.
+match_did(ToDID) ->
+    case gen_server:call(?MODULE, {'match_did', kz_util:to_list(ToDID)}) of
+        {'error', 'not_found'} ->
+            lager:debug("failed to find rate for ~s", [ToDID]),
+            {'ok', []};
+        {'ok', {_Prefix, RateIds}} ->
+            lager:info("candidate rates for ~s: ~s ~p", [ToDID, _Prefix, RateIds]),
+            load_rates(RateIds)
+    end.
+
+-spec load_rates(ne_binaries()) -> {'ok', kz_json:objects()}.
+load_rates(RateIds) ->
+    {'ok', lists:foldl(fun load_rate/2, [], RateIds)}.
+
+-spec load_rate(ne_binary(), kz_json:objects()) -> kz_json:objects().
+load_rate(RateId, Acc) ->
+    case kz_datamgr:open_cache_doc(?KZ_RATES_DB, RateId) of
+        {'ok', RateDoc} -> [RateDoc | Acc];
+        {'error', _} -> Acc
+    end.
+
+-spec init([]) -> {'ok', state()}.
+init([]) ->
+    PidRef = spawn_monitor(?MODULE, 'build_trie', [self()]),
```

Comment:
The plan is to have the trie response {error, not_ready} while the trie is building. I also didn't want all the garbage involved with generating the trie in the main trie server process, hence the spawned worker.

A longer term plan is to try the trie and, on error, try the couch view (if trie is enabled). So even if a rate request comes in the first couple seconds of the app being started, it can still be serviced, albeit by the slower view lookup.

I like monitors over links when the spawning process doesn't care about the spawned process' health. The code can be changed to detect the DOWN when state is still the {pid(), reference()} and restart the trie building, for instance. I feel like monitors are lighter than links but that's entirely unsubstantiated.